astroplan/constraints.py

```diff
@@ -0,0 +1,577 @@
+
+    def _compute_constraint(self, time_range, observer, targets,
+                            time_resolution=DEFAULT_TIME_RESOLUTION):
+        cached_altaz = self._get_altaz(time_range, observer, targets,
+                                       time_resolution=time_resolution)
+        altaz = cached_altaz['altaz']
+        if self.min is None and self.max is not None:
+            mask = altaz.secz < self.max
+        elif self.max is None and self.min is not None:
+            mask = self.min < altaz.secz
+        elif self.min is not None and self.max is not None:
+            mask = (self.min < altaz.secz) & (altaz.secz < self.max)
+        else:
+            raise ValueError("No max and/or min specified in "
+                             "AirmassConstraint.")
+        return mask
+
+class AtNightConstraint(Constraint):
+    """
+    Constrain the Sun to be below ``horizon``.
+    """
+    @u.quantity_input(horizon=u.deg)
+    def __init__(self, max_solar_altitude=0*u.deg, force_pressure_zero=True):
+        """
+        Parameters
+        ----------
+        max_solar_altitude : `~astropy.units.Quantity`
+            Define "night" as when the sun is below ``max_solar_altitude``.
+            Default is zero degrees altitude.
+
+        force_pressure_zero : bool (optional)
+            Force the pressure to zero for solar altitude calculations. This
+            avoids errors in the altitude of the Sun that can occur when the
+            Sun is below the horizon and the corrections for atmospheric
+            refraction return nonsense values. Default is `True`.
+        """
+        self.max_solar_altitude = max_solar_altitude
+        self.force_pressure_zero = force_pressure_zero
+
+    @classmethod
+    def twilight_civil(cls, **kwargs):
+        """
+        Consider nighttime as time between civil twilights (-6 degrees).
+        """
+        return cls(max_solar_altitude=-6*u.deg, **kwargs)
+
+    @classmethod
+    def twilight_nautical(cls, **kwargs):
+        """
+        Consider nighttime as time between nautical twilights (-12 degrees).
+        """
+        return cls(max_solar_altitude=-12*u.deg, **kwargs)
+
+    @classmethod
+    def twilight_astronomical(cls, **kwargs):
+        """
+        Consider nighttime as time between astronomical twilights (-18 degrees).
+        """
+        return cls(max_solar_altitude=-18*u.deg, **kwargs)
+
+    def _get_solar_altitudes(self, time_range, observer, targets,
+                             time_resolution=DEFAULT_TIME_RESOLUTION):
+        if not hasattr(observer, '_altaz_cache'):
+            observer._altaz_cache = {}
+
+        times = time_grid_from_range(time_range,
+                                     time_resolution=time_resolution)
+
+        aakey = (tuple(times.jd), 'sun')
+
+        if aakey not in observer._altaz_cache:
+            if self.force_pressure_zero:
+                observer_old_pressure = observer.pressure
+                observer.pressure = 0
```

Comment:
Ditto.  This form of stateful, temporary overrides is prone to problems when there are intervening exceptions before the restoration of the state.  e.g. someone working in an interactive session with an `observer` they won't realize that the pressure that they carefully set when constructing the object has been zeroed.